File: Recursion/51H. N-Queens.py
'''
Method:
1. Create a board. board = [list('.'*n) for i in range(n)]
   Note: we convert a string to a list, because the string doesn't support item assignment in Python.
2. Save a valid solution.
   Res = []
   for r in board:
       Res.append(''.join(deepcopy(r)))
   res.append(Res)
   Note: need to first deepcopy the result, otherwise it will be kept modifying the element values
         also need to convert the list back to the string. Using ''.join(list).
'''
class Solution(object):
    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        res = []
        board = [list('.'*n) for i in range(n)]
        def isValid(board, row, col):
            for i in range(len(board)):
                if board[i][col]=='Q':
                    return False
            i, j = row-1, col+1
            while i>=0 and j<len(board):
                if board[i][j]=='Q':
                    return False
                i -= 1
                j += 1
            i, j = row-1, col-1
            while i>=0 and j>=0:
                if board[i][j]=='Q':
                    return False
                i -= 1
                j -= 1
            return True
        
        def backtrack(board, row):
            if row == len(board):
                Res = []
                for r in board:
                    Res.append(''.join(deepcopy(r)))
                res.append(Res)
                return
            for col in range(len(board[0])):
                if isValid(board, row, col):
                    board[row][col] = 'Q'
                    backtrack(board, row+1)
                    board[row][col] = '.'
        
        backtrack(board, 0)
        return res

File: Recursion/51H. N-Queens.cpp
/*
51H. N-Queens
Backtracking

Description:
The n-queens puzzle is the problem of placing n queens on an n x n chessboard such that no two queens attack each other.
Given an integer n, return all distinct solutions to the n-queens puzzle.
Each solution contains a distinct board configuration of the n-queens' placement, 
where 'Q' and '.' both indicate a queen and an empty space, respectively.

Example 1:
Input: n = 4
Output: [[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]]
Explanation: There exist two distinct solutions to the 4-queens puzzle as shown above

Example 2:
Input: n = 1
Output: [["Q"]]

Constraints:
1 <= n <= 9

Similar Questions:
N-Queens II - Hard
Grid Illumination - Hard
*/

/*
Method: Create a vector<string> to hold the layout of the board.
        Then we check the board row by row.
        In each row, we also check each column and also check if the cell is a valid position to place the queen.
        Once we place the queen, then we call the recursion to check the following rows.
        Once it is done, we take out the queen from the current cell and move to the next one.
*/
class Solution {
public:
    vector<vector<string>> res;
    vector<vector<string>> solveNQueens(int n) {
        vector<string> board(n, string(n, '.'));
        backtrack(board, 0);
        return res;
    }
    void backtrack(vector<string>& board, int row){
        if(row == board.size()){
            res.push_back(board);
            return;
        }
        int n = board.size();
        for(int col=0; col < n; col++){
            if(is_valid(board, row, col)){
                board[row][col] = 'Q';
                backtrack(board, row+1);
                board[row][col] = '.';
            }
        }
    }
    bool is_valid(vector<string>& board, int row, int col){
        // Check if there is queen in the same column
        for(int i=0; i<board.size(); i++){
            if(board[i][col]=='Q'){
                return false;
            }
        }
        // Check upper right side
        for(int i=row-1, j=col+1; i>=0 && j<board.size(); i--, j++){
            if(board[i][j]=='Q'){
                return false;
            }
        }
        // Check upper left side
        for(int i=row-1, j=col-1; i>=0 && j>=0; i--, j--){
            if(board[i][j]=='Q'){
                return false;
            }
        }
        return true;
    }
};
